Bugsnag/Helpers/BSGAppHangDetector.m

```diff
@@ -57,98 +53,147 @@ - (void)startWithDelegate:(id<BSGAppHangDetectorDelegate>)delegate {
-                            thread.errorReportingThread = idx == 0;
-                        }];
-                    } else {
-                        threads = BSGArrayWithObject([BugsnagThread mainThread]);
-                    }
-                    
-                    __strong typeof(weakDelegate) strongDelegate = weakDelegate;
-                    
-                    [strongDelegate appHangDetectedAtDate:date withThreads:threads systemInfo:systemInfo];
-                    
-                    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
-                    bsg_log_info("App hang has ended");
-                    
-                    [strongDelegate appHangEnded];
-                }
-            });
+            self.processingDeadline = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(threshold * NSEC_PER_SEC));
+            dispatch_semaphore_signal(self.processingStarted);
+            isProcessing = YES;
+            return;
         }
         
-        // "Inside the event processing loop before the run loop sleeps, waiting for a source or timer to fire"
         if (activity == kCFRunLoopBeforeWaiting) {
-            if (semaphore) {
-                dispatch_semaphore_signal(semaphore);
+            if (isProcessing) {
+                dispatch_semaphore_signal(self.processingFinished);
+                isProcessing = NO;
             }
+            return;
         }
     };
     
-    // A high `order` is required to ensure our observer runs after others that may introduce an app hang.
+    // A high `order` is required to ensure our kCFRunLoopBeforeWaiting observer runs after others that may introduce an app hang.
     // Once such culprit is -[UITableView tableView:didSelectRowAtIndexPath:] which is run in a
     // _afterCACommitHandler, which is invoked via a CFRunLoopObserver.
     CFIndex order = INT_MAX;
-    self.observer = CFRunLoopObserverCreateWithHandler(NULL, kCFRunLoopAfterWaiting | kCFRunLoopBeforeWaiting, true, order, observerBlock);
-    
-    CFRunLoopMode runLoopMode = CFRunLoopCopyCurrentMode(CFRunLoopGetCurrent());
-    // The run loop mode will be NULL if called before the run loop has started; e.g. in a +load method.
-    if (runLoopMode) {
-        // If we are already in the run loop (e.g. in app delegate) start monitoring immediately so that app hangs during app launch are detected.
-        observerBlock(self.observer, kCFRunLoopAfterWaiting);
-        CFRelease(runLoopMode);
-    }
+    CFRunLoopActivity activities = kCFRunLoopAfterWaiting | kCFRunLoopBeforeSources | kCFRunLoopBeforeWaiting;
+    self.observer = CFRunLoopObserverCreateWithHandler(NULL, activities, true, order, observerBlock);
+    
+    // Start monitoring immediately so that app hangs during launch can be detected.
+    // Note that because scene-based apps start in UIApplicationStateBackground, hangs in
+    // -[AppDelegate application:didFinishLaunchingWithOptions:] will be ignored.
```

Comment:
Is there any way we can detect this in future?

Reply:
We could temporarily allow background hangs to be acted upon, we'd need to find a suitable point at which to reenable the usual logic though.

Briefly discussed this with Tom and agreed it was better to address it as a separate issue to be discussed with the product team.